### PR TITLE
Add a hook to allow world providers to change the lightmap colours (1.12)

### PR DESCRIPTION
**Pull request by bs2609**
1.12 version of #3922.

### CODE REVIEW

**Comment by mezz** [comment]
Looks good to me and approved by Fry.

File: patches/minecraft/net/minecraft/world/WorldProvider.java.patch [hunk 1]
```diff
@@ -152,6 +152,19 @@
...
++     * Will be called for each 16*16 combination of sky/block light values.
++     *
++     * @param sunBrightness Current sun brightness.
++     * @param skyLight Sky light brightness factor.
++     * @param blockLight Block light brightness factor.
++     * @param colors The color values that will be used: [r, g, b].
++     *
++     * @see net.minecraft.client.renderer.EntityRenderer#updateLightmap(float)
++     */
++    public void getLightmapColors(float sunBrightness, float skyLight, float blockLight, float[] colors) {}
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/world/WorldProvider.java.patch]: Why byref vs return?

Comment (bs2609) [reply patches/minecraft/net/minecraft/world/WorldProvider.java.patch]: So as to be sure it's a nonnull array of size 3, to save a check where it's used.

Comment (howtonotwin) [reply patches/minecraft/net/minecraft/world/WorldProvider.java.patch]: Could also return `@Nonnull Triple<Float, Float, Float>`.

Comment (williewillus) [reply patches/minecraft/net/minecraft/world/WorldProvider.java.patch]: @howtonotwin that's strictly less performant because the floats are boxed

Comment (liach) [reply patches/minecraft/net/minecraft/world/WorldProvider.java.patch]: And you need to create another triple to modify them

**Comment by bs2609** [comment]
Added `partialTicks` as a parameter, to allow for effects that can change with time independently of light level.

Example implementation:
```java
public void getLightmapColors(float partialTicks, float sunBrightness, float skyLight, float blockLight, float[] colors)
{
	final int cycleTime = 60;

	long tick = world.getTotalWorldTime();
	float hue = (tick + partialTicks) % cycleTime / cycleTime;

	int rgb = java.awt.Color.HSBtoRGB(hue, 1.0f, 1.0f);

	float r = (rgb >> 16 & 255) / 255f;
	float g = (rgb >>  8 & 255) / 255f;
	float b = (rgb       & 255) / 255f;

	colors[0] = (colors[0] + r) / 2f;
	colors[1] = (colors[1] + g) / 2f;
	colors[2] = (colors[2] + b) / 2f;
}
```

**Pull request merged**